### PR TITLE
[BugFix] Attempt at better redirecting stdout to file.

### LINKED ISSUE #8975: [Help wanted!!]When outputs redirect from terminal to a log file, Progress Bar will output in new line but update original line
## 🐛 Bug

Bellow is my log file looks like that full of useless informatio.
image

### Expected behavior

**Dynamic refresh in one line or just print once every epoch end.**

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.4.1
- PyTorch Version (e.g., 1.8) 1.9.0
- Python version: python 3.8
- OS (e.g., Linux):
- CUDA/cuDNN version: 
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for file being provided to `ProgressBar` and adapt tqdm to avoid over-flowing the logs.

Fixes #8975

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/progress.py [hunk 1]
```diff
@@ -43,8 +42,36 @@ class tqdm(_tqdm):
...
+                    ]
+                    self.fp.seek(0)
+                    self.fp.truncate()
+                    for line in lines:
+                        line = line.replace("[A", "")
+                        self.fp.write(line)
+        else:
+            msg = self.__repr__()
+            super().display(msg=msg, pos=0)
+
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/progress.py:73]: is this a fix that could be upstreamed to `tqdm` directly? what guarantees that this override stays in sync with `tqdm`'s API over time?

File: pytorch_lightning/callbacks/progress.py [hunk 2]
```diff
@@ -43,8 +42,36 @@ class tqdm(_tqdm):
     Custom tqdm progressbar where we append 0 to floating points/strings to prevent the progress bar from flickering
     """
 
-    @staticmethod
```
Comment (carmocca) [inline pytorch_lightning/callbacks/progress.py:46]: You removed the `staticmethod` annotation

Comment (tchaton) [reply pytorch_lightning/callbacks/progress.py:46]: added back !

File: pytorch_lightning/callbacks/progress.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/progress.py:73]: No I don't believe so. 

I would call this more a hack to prevent massive logs while re-directing to a file, but I believe there might smarter way to do this.

Concerning the API, I would say tqdm API is pretty stable right now and this hack is just overriding their display hook.

I am pretty confident they won't be divergence there.

File: pytorch_lightning/callbacks/progress.py [hunk 3]
```diff
@@ -274,10 +302,11 @@ def init_validation_tqdm(self):
 
     """
 
-    def __init__(self, refresh_rate: int = 1, process_position: int = 0):
+    def __init__(self, refresh_rate: int = 1, process_position: int = 0, file=sys.stdout):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress.py:305]: probably need to add a typehint for the stream

File: tests/callbacks/test_progress_bar.py [hunk 4]
```diff
@@ -558,3 +558,29 @@ def _test_progress_bar_max_val_check_interval(
     total_val_batches = total_val_batches * val_checks_per_epoch
     if trainer.is_global_zero:
         assert trainer.progress_bar_callback.main_progress_bar.total == total_train_batches + total_val_batches
+
+
+def test_progress_bar_redirect(tmpdir):
+
+    """
+    This test ensures the log saved to a file are being trimmed.
+    """
```
Comment (awaelchli) [inline tests/callbacks/test_progress_bar.py:567]: can we agree on docstring formatting 😅 ? I'm seeing this new pattern emerge recently but why? I thought we had blacken-docs and all that fancy formatting.

File: tests/callbacks/test_progress_bar.py [hunk 5]
```diff
@@ -558,3 +558,29 @@ def _test_progress_bar_max_val_check_interval(
...
+
+    log_file = tmpdir / "tmp.log"
+
+    with open(log_file, "w") as fw:
+        trainer = Trainer(max_epochs=4, callbacks=ProgressBar(file=fw))
+        trainer.fit(LoggingModel())
+
+    with open(log_file) as fr:
+        contents = fr.readlines()
+        assert len(contents) == 10
```
Comment (awaelchli) [inline tests/callbacks/test_progress_bar.py:585]: why are there 10 lines? 
I guess the sanity check bar is also included right?

File: tests/callbacks/test_progress_bar.py [hunk 6]
```diff
@@ -558,3 +558,29 @@ def _test_progress_bar_max_val_check_interval(
...
+    class LoggingModel(BoringModel):
+        def validation_step(self, batch, batch_idx):
+            loss = super().validation_step(batch, batch_idx)
+            self.log("val_loss", loss["x"], prog_bar=True)
+            return loss
+
+    log_file = tmpdir / "tmp.log"
+
+    with open(log_file, "w") as fw:
+        trainer = Trainer(max_epochs=4, callbacks=ProgressBar(file=fw))
```
Comment (awaelchli) [inline tests/callbacks/test_progress_bar.py:580]:
```suggestion
        trainer = Trainer(default_root_dir=tmpdir, max_epochs=4, callbacks=ProgressBar(file=fw))
```

File: tests/callbacks/test_progress_bar.py (hunk 4 above)
Comment (carmocca) [reply tests/callbacks/test_progress_bar.py:567]: AFAIK `blacken-docs` is just for Sphinx docs.

@Borda opened https://github.com/PyTorchLightning/pytorch-lightning/pull/8584 for docstring auto-formatting. Not sure if it would fix the extra space before

File: pytorch_lightning/callbacks/progress.py [hunk 7]
```diff
@@ -43,6 +42,35 @@ class tqdm(_tqdm):
     Custom tqdm progressbar where we append 0 to floating points/strings to prevent the progress bar from flickering
     """
 
+    def __init__(self, *args, **kwargs):
+        self.using_text_io_wrapper = False
+        super().__init__(*args, **kwargs)
+
+        self.using_text_io_wrapper = hasattr(self, "fp") and isinstance(self.fp, io.TextIOWrapper)
+
+    def display(self, msg=None, pos=None):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/progress.py:51]: Can you explain a bit what this does and how it does it?

Why do we need all the if/else instead of:

```python
super().display(msg=msg, pos=pos)
if self.using_textio_wrapper and self.n == self.total:
    fix_file_lines()
```

File: tests/callbacks/test_progress_bar.py (hunk 4 above)
Comment (tchaton) [reply tests/callbacks/test_progress_bar.py:567]: Yes, cleaned out !

File: tests/callbacks/test_progress_bar.py (hunk 6 above)
Comment (tchaton) [reply tests/callbacks/test_progress_bar.py:580]: Thanks !

**Comment by tchaton** [comment]
Closing this PR as this is too much of a hack.
